static-build/posts/2021/09/cors/index.md

```diff
@@ -0,0 +1,354 @@
+- Firefox and Safari go a step further, and try to fully isolate sites, although how this works is currently pretty different between the two.
+
+## The same-origin policy
+
+Back in 1995, Netscape 2 landed with two amazing new features: LiveScript (you probably know this better as 'JavaScript'), and HTML frames. Frames let you embed one page in another, and LiveScript could interact with both pages.
+
+Netscape realised that this presented a security issue, so they decided that cross-frame scripting would only be allowed if both pages had the same _origin_.
+
+<style>
+.origin-example {
+  display: grid;
+  grid-template-columns: max-content max-content;
+}
+.origin {
+  grid-area: 1 / 1;
+  color: #ff824a;
+}
+.path {
+  grid-area: 1 / 2;
+}
+.origin-label {
+  grid-area: 2 / 1;
+  text-align: center;
+  display: grid;
+  grid-template-columns: 1fr max-content 1fr;
+  align-items: top;
+  gap: 0.3em;
+  color: #ff824a;
+}
+.origin-label::before,
+.origin-label::after {
+  content: '';
+  border: 1px solid currentcolor;
+  height: 50%;
+}
+.origin-label::before {
+  border-width: 0 0 1px 1px;
+}
+.origin-label::after {
+  border-width: 0 1px 1px 0;
+}
+</style>
+
+<div class="code-example"><pre class="language-xml"><code><div class="origin-example"><div aria-labelledby="origin-label" class="origin">https://jakearchibald.com:443</div><div class="path">/2021/blah/?foo#bar</div><div id="origin-label" class="origin-label">The origin</div></div></code></pre></div>
+
+The idea was that sites on the same origin are more likely to have the same owner. That wasn't completely true, since a lot of sites divided content by URLs such as `http://example.com/~jakearchibald/`, but the line had to be drawn somewhere.
+
+From that point, features that granted deep visibility into a resource were limited to same-origin. This included `new ActiveXObject('Microsoft.XMLHTTP')` which first appeared in IE5 in 1999, and later became the web standard `XMLHttpRequest`.
+
+## Origins vs sites
+
+Some web features don't deal with origins, they deal with 'sites'. For instance, `https://help.yourbank.com` and `https://profile.yourbank.com` are different origins, but they're the same site. Cookies are the most common feature that operate at a site level, as you can create cookies that are sent to all subdomains of `yourbank.com`.
+
+But how does the browser know that `https://help.yourbank.com` and `https://profile.yourbank.com` are part of the same site, but `https://yourbank.co.uk` and `https://jakearchibald.co.uk` are different sites? I mean… they all have three parts separated by dots.
+
+Well, the answer was a bunch of heuristics in each browser, but in 2007 Mozilla swapped their heuristics for a list. That list is now maintained as a separate community project known as the [public suffix list](https://publicsuffix.org/), and it's used by all browsers and many other projects.
```

Comment:
I found the tool https://site-and-origin.securitum.com/ helpful for playing with this. Consider linking out to it.

Reply:
I've added an inline example that does the same thing (although I need to tweak it a bit so it handles IP addresses)